lib/logging/logging.dart

```diff
@@ -248,30 +316,128 @@ class LoggingScreen extends Screen {
   }
 }
 
-bool _isNotNull(dynamic serviceRef) {
-  return serviceRef != null && serviceRef['kind'] != 'Null';
+/// Receive and log stdout / stderr events from the VM.
+///
+/// This class buffers the events for up to 1ms. This is in order to combine a
+/// stdout message and its newline. Currently, `foo\n` is sent as two VM events;
+/// we wait for up to 1ms when we get the `foo` event, to see if the next event
+/// is a single newline. If so, we add the newline to the previous log message.
+class _StdoutEventHandler {
+  final LoggingScreen loggingScreen;
+  final String name;
+  final bool isError;
+
+  LogData buffer;
+  Timer timer;
+
+  _StdoutEventHandler(this.loggingScreen, this.name, {this.isError = false});
+
+  void handle(Event e) {
+    final String message = decodeBase64(e.bytes);
+
+    if (buffer != null) {
+      timer?.cancel();
+
+      if (message == '\n') {
+        loggingScreen._log(new LogData(
+          buffer.kind,
+          buffer.details + message.replaceAll('\n', r'\n'),
+          buffer.timestamp,
+          summary: buffer.summary + message.replaceAll('\n', r'\n'),
+          isError: buffer.isError,
+        ));
+        buffer = null;
+        return;
+      }
+
+      loggingScreen._log(buffer);
+      buffer = null;
+    }
+
+    String summary = message;
+    if (message.length > 200) {
+      summary = message.substring(0, 200) + '…';
+    }
+    summary = summary.replaceAll('\t', r'\t');
```

Comment:
What's the reason for these replacements? Won't replacing characters like this (for ex.) make an actual newline and the literal characters `\n` in the users log render the same in this view? This might complicate debugging, for ex. if your app showing a literal `\n` where you expected a newline?

Reply:
That's a good point - it would be hard to differentiate between a newline and a sequence of two `\n` chars.

Without this however, you won't know if you have whitespace chars in your output - they won't show in the rendering.

I'm not sure if there's a solution that covers both situations; if no, we likely want to make the whitespace chars visible.

Reply:
> Without this however, you won't know if you have whitespace chars in your output - they won't show in the rendering

Do you mean because they collapse in HTML? If so, don't the pre/whitespace settings help with that?

My concern here was that any manipulation of the data here could lose data - it'd be better to keep the raw string and deal with anything like this in the rendering as-appropriate? Fox ex. It might be important that the user can copy/paste an item from the log and have it as was printed verbatim?

Reply:
> Do you mean because they collapse in HTML? If so, don't the pre/whitespace settings help with that?

It would with the formatting aspect, but not w/ knowing whether it was a tab, several spaces, ...

I can remove this part and we can re-assess if this ever comes up as an issue.

Reply:
> but not w/ knowing whether it was a tab, several spaces

Oh, gotcha! I guess this is no different to when the user `prints` and sees it written anywhere else so probably people are used to selecting/inspecting/copy&pasting to review, but we could consider a button that toggles rendering whitespace in the details?

Reply:
> we could consider a button that toggles rendering whitespace in the details

That sounds great - a good solution. It could be a raw data vs a rendering option, esp. for some log entries where we interpret the data (frame events, gc events, ...).

Reply:
👍 I added a note to the doc so this doesn't get forgotten (though we can start using issues here if you think makes more sense).

Reply:
We should likely move the work we think we're likely to do from the doc into the issue tracker.